include/swift/AST/TypeTransform.h

```diff
@@ -939,7 +940,8 @@ case TypeKind::Id:
           });
 
         if (didRemoveLifetimeDependencies) {
-          extInfo = extInfo->withLifetimeDependencies(substDependenceInfos);
+          extInfo = extInfo->withLifetimeDependencies(
+              ctx.AllocateCopy(substDependenceInfos));
```

Comment:
Maybe it's best to sink the AllocateCopy down into withLifetimeDependencies()

Reply:
Is there some rule for when `withLifetimeDependencies` needs its own allocation? Could that at least be commented in the API?

We call this in several other places and I don't know which need an allocation and which don't. e.g.
`SILFunctionType::getSubstLifetimeDependencies`

Reply:
It needs its own allocation if the LifetimeDependencies that you pass in are themselves not in the global AST context arena. It's safest to move the AllocateCopy into withLifetimeDependencies() itself, so that callers don't have to worry about creating an ExtInfo with a dangling pointer in it.

Reply:
I did not originally want to do this because `withLifetimeDependencies ` is called from places with an already arena allocated pointer, and proactively allocating within it would not allow for sharing.

Reply:
perhaps the LifetimeDependencies objects should always be allocated and uniqued by the ASTContext then, so you'll pass them around by pointer and the pointer will always have global extent.

Reply:
I was hoping to use something akin to `llvm::BumpPtrAllocatorImpl::identifyObject` to ensure the passed input arg is safe in assert builds. Not everything is plumbed through to try that out. 

I added a C++ delete overload for `withLifetimeDependencies` when input argument is `SmallVectorImpl` for now.

Reply:
> perhaps the LifetimeDependencies objects should always be allocated and uniqued by the ASTContext then

This sounds promising.

Reply:
I created rdar://151811669 (LifetimeDependencies objects should always be allocated and uniqued by the ASTContext)